# "Pre-lecture" HW1
## Question 1

In [5]:
import pandas as pd

# URL of the Titanic dataset
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"

# Load the dataset into a pandas DataFrame
titanic_data = pd.read_csv(url)

# Display the first few rows of the dataset
titanic_data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [11]:
# Load the dataset into a pandas DataFrame
titanic_data = pd.read_csv(url)

# Count missing values in each column
missing_values = titanic_data.isnull().sum()

# Display the count of missing values for each column
print(missing_values)

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


In [10]:
# Load the dataset into a pandas DataFrame
titanic_data = pd.read_csv(url)

# Count missing values in each column and sum them to get the total
total_missing_values = titanic_data.isnull().sum().sum()

# Display the total number of missing values
print(f"Total missing values in the dataset: {total_missing_values}")

Total missing values in the dataset: 866


### Summary of Interactions with ChatGPT

1. Request for an Interesting Dataset
    I asked ChatGPT to find an amusing, funny, or otherwise interesting dataset that:
    - Had missing values,
    - Was available online as a CSV file.
     
     ChatGPT suggested the Titanic dataset, which is a well-known dataset used for educational purposes. It contains missing values in columns such as "Age," "Cabin," and "Embarked." ChatGPT provided the link to access the dataset: [Titanic Dataset CSV](https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv).
    
    
2. Importing the Dataset
    I requested the code to import the Titanic dataset and display the count of missing values. ChatGPT provided Python code using the `pandas` library to:
    - Import the dataset
    - Count the number of missing values in each column using the `isnull().sum()` method.
   
   Here's the code provided:

In [ ]:
import pandas as pd

# URL of the Titanic dataset
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"

# Load the dataset into a pandas DataFrame
titanic_data = pd.read_csv(url)

# Count missing values in each column
missing_values = titanic_data.isnull().sum()

# Display the count of missing values for each column
print(missing_values)

3. Displaying the Dataset
    When I asked for the code to import and display the dataset, ChatGPT provided an additional snippet to show the first few rows using `head()`:

In [ ]:
print(titanic_data.head())

4. Counting the Total Missing Values
    I asked how to calculate the total number of missing values across the entire dataset. ChatGPT provided a code snippet that summed the missing values in all columns:

In [ ]:
# Count missing values in each column and sum them to get the total
total_missing_values = titanic_data.isnull().sum().sum()

# Display the total number of missing values
print(f"Total missing values in the dataset: {total_missing_values}")

    This code returns the total number of missing values across the entire dataset.

[Chatbot transcript](https://chatgpt.com/share/bccfa142-af26-4b0b-a588-12dd0b6d0163)

## Question 2

In [9]:
import pandas as pd

url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv" # we are still using the dataset in Q1
df = pd.read_csv(url)
rows, columns = df.shape
print(f"The DataFrame has {rows} rows and {columns} columns.")

The DataFrame has 891 rows and 12 columns.


- Observations (Rows) are the individual entries of the dataset. Rach observations represents a single unit of data. 

- Variables (Columns) are the traits that describe each observation. They represent the characteristics of what you are observing.

### Summary of Interactions with ChatGPT

1. Dataset Reference:

    The Titanic dataset was provided via the following link: [Titanic Dataset CSV](https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv).
    

2. Determining the Number of Rows and Columns in a DataFrame:
    - You were shown how to read the dataset into a pandas DataFrame using `pd.read_csv()`.
    - To determine the number of rows and columns, you can use the `shape` attribute of a DataFrame:

In [ ]:
rows, columns = df.shape
print(f"The DataFrame has {rows} rows and {columns} columns.")

    - This provides the total number of observations (rows) and variables (columns).
    
3. Explanation of Observations and Variables:
    - Observations: Individual entries in the dataset. In this case, each observation represents a single passenger on the Titanic.
    
    - Variables: The attributes or characteristics of the observations. These are represented by the columns in the dataset, such as `Name`, `Age`, `Fare`, `Pclass`, and `Survived`, among others.

[Chatbot transcript](https://chatgpt.com/share/19925e12-ec4d-4f0f-bd88-aae676cae2c8)

## Question 3

In [10]:
# Get a summary of numerical columns
numerical_summary = df.describe()
print("Numerical Columns Summary:")
numerical_summary

Numerical Columns Summary:


PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [13]:
# Examples of categorical columns

# Summary for 'Sex' column
sex_counts = df['Sex'].value_counts()
print("\nSex Distribution:")
print(sex_counts)


Sex Distribution:
Sex
male      577
female    314
Name: count, dtype: int64


In [14]:
# Summary for 'Embarked' column
embarked_counts = df['Embarked'].value_counts()
print("\nEmbarked Distribution:")
print(embarked_counts)


Embarked Distribution:
Embarked
S    644
C    168
Q     77
Name: count, dtype: int64


In [15]:
# General information about the DataFrame
general_info = df.info()
print("\nGeneral Data Overview:")
print(general_info)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB

General Data Overview:
None


### Summary of Interactions with ChatGPT

1. How to provide simple summaries of dataset columns:
    
    - You asked how to provide simple summaries of the columns in a dataset.
    - I explained how to generate these summaries, particularly focusing on numerical and categorical columns. For numerical columns, I mentioned using `.describe()` to get statistics like count, mean, standard deviation, min, max, and quartiles. For categorical columns, I suggested using `.value_counts()` to see the distribution of categories.
    
2. Summaries for Titanic dataset:
    - You provided a link to the Titanic dataset and requested the simplest kinds of summaries.
    - I provided Python code to load the dataset and generate simple summaries:
        - For numerical columns, the use of `.describe()` provides count, mean, standard deviation, min, max, and quartiles.
        - For categorical columns, I showed how to use `.value_counts()` for columns like 'Sex', 'Embarked', and 'Pclass'.
        - Additionally, I provided code to use `.info()` for a general overview of the dataset, which includes non-null counts and data types.

[Chatbot transcript](https://chatgpt.com/share/f944a75b-55e6-4a0e-84d0-09d515c5ac5d)

## Question 4

`df.shape` provides the total number of rows and columns including the missing values, in the Titanic dataset which is 891 rows and 12 columns.

`df.describe()` provides analysis of only numeric columns, which are non-null values. There are 7 numeric columns in total for the Titanic dataset, and 891 rows for most of the columns except for 'Age', it only has 714 rows by looking at the count. 

- There exists discrepancy between the results of columns, since `df.describe()` only includes numeric columns (int, float). It does not analyze columns with data types like object and columns with all missing values, while `df.shape` includes them and gives the total number of columns. Therefore if the dataset contains non-numeric or null values columns, they will not appear in the `df.describe()` output. For the rows, when applying `df.describe()`, 'count' gives the imformation of the number of non-missing values for each column. `df. shape` give the total number of rows, including the missing values. So the discrepancy here would be caused by the missing values. 


- In the case of Titanic dataset, there are 12 columns in total (concluded by `df.shape`) and by applying `df.info()`, we know the column's types and by comparing these imformation with the result of `df.describe()` (7 numeric columns), we learn that the discrepancy is due to the object columns, which are not included in the result given by `df.describe()`. There are also discrepancy in rows for certain columns such as 'Age'. It has 714 rows according to `df.describe()`, `df.shape` shows it has 891 rows, since it included the rows with missing value. By using `.value_counts(dropna=False)`, we can see how many missing values the column has.


In [3]:
missing_values = df['Age'].value_counts(dropna=False)
print(missing_values)

Age
NaN      177
24.00     30
22.00     27
18.00     26
28.00     25
        ... 
36.50      1
55.50      1
0.92       1
23.50      1
74.00      1
Name: count, Length: 89, dtype: int64


Here, we can see that there are 177 missing values in the 'Age' column, and when you add 714 of non-missing values with 177 missing values, the total is 891 which is the same with the total number of rows given by `df.shape`.

### Summary of Interactions with ChatGPT

1. Dataset Overview:
   - You provided the link to the [Titanic Dataset CSV](https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv).
   
   - I explained that the dataset typically contains columns like `PassengerId`, `Pclass`, `Name`, `Sex`, `Age`, `SibSp`, `Parch`, `Ticket`, `Fare`, `Embarked`, and `Survived`.

2. Dataset Shape and Summary Statistics:
    - I described how you can use `df.shape` to get the size of the dataset (number of rows and columns) and `df.describe()` to get summary statistics for numeric columns.

    - I also clarified that `df.describe()` analyzes only numeric columns, providing statistics like `count`, `mean`, `std`, `min`, `max`, etc.
    
    - I explained that the `count` value in `df.describe()` reflects the number of non-null values for each numeric column, which can be less than the total number of rows due to missing data.
    
3. Discrepancies Between `df.shape` and `df.describe()`:
    - I explained that the discrepancy in columns arises because `df.shape` includes all columns, both numeric and non-numeric, while `df.describe()` only analyzes numeric columns.
    
    - Non-numeric columns like `Name`, `Sex`, `Ticket`, `Cabin`, and `Embarked` are excluded from `df.describe()` because they are not suitable for numerical summary statistics.
    
4. Handling Missing Values in Summary Statistics:
    - I pointed out that missing values in numeric columns can cause discrepancies between the total number of rows and the count reported by `df.describe()`, as it only counts non-null values.
    
5. Discrepancy Between `df.shape` and `df.info()`:
    - You observed that `df.info()` showed 11 columns, while `df.shape` indicated 12.
    
    - I explained that this discrepancy could be due to various factors such as columns with all missing values, hidden columns, or columns with problematic data types that might not be included in `df.info()`.

    - I provided suggestions for diagnosing the issue, including checking column names, data types, and missing values.

[Chatbot transcript](https://chatgpt.com/share/3b61f37a-d620-4c1a-a551-e095613ffb0f)

## Question 5

An attribute is a property or value linked to an object that provides information about it. It does not require parentheses when accessed.

   - e.g. `df.shape` gives the dimensions of a dataframe

A method is a function tied to an object that performs an action. It requires parentheses when called, as it may take arguments.
   
   - e.g. `df.describe()` computes and returns a summary of measures for the numerical columns of a dataframe.

### Summary of Interactions with ChatGPT

1. Question about Attributes vs. Methods in Python:
    - You asked for the difference between an "attribute" and a "method" in the context of a DataFrame in pandas.
    - I explained that:
        - Attributes (e.g., `df.shape`) are values or properties tied to an object and do not require parentheses. They provide information but do not perform any actions.
        - Methods (e.g., `df.describe()`) are functions associated with an object that perform actions or operations. They require parentheses because they may take arguments and execute code.
        

[Chatbot transcript](https://chatgpt.com/share/74ae1400-e63a-466d-8fe9-6cd35faef639)

# "Post-lecture" HW

## Question 6

The statistics provided by `.describe()` include:

    - Count: The number of non-missing values for each column.
    
    - Mean: The average value of the entries.
    
    - Standard Deviation (std): A measure of the spread of values. It quantifies how much the values typically deviate from the mean. 
    
    - Min (Minimum): The smallest value in the column.
    
    - 25% (First Quartile, Q1): The value below which 25% of the data falls. Also known as the first quartile, it marks the point below which the lowest 25% of the data lies.
    
    - 50% (Median, Q2): The middle value of the data when arranged in ascending order. Also known as the second quartile, or the median, it splits the data into two equal halves.
    
    - 75% (Third Quartile, Q3): The value below which 75% of the data falls. Also known as the third quartile, it marks the point below which the lowest 75% of the data lies.
    
    - Max (Maximum): The largest value in the column.

### Summary of Interactions with ChatGPT

1. Discussion on df.describe() in Pandas:
    - You asked about the definitions of the summary statistics provided by the df.describe() method in pandas.

    - I explained the definitions of the following statistics:
        - Count: The number of non-missing values in a column.
        - Mean: The average of the values in the column.
        - Standard Deviation (std): The measure of the variation or spread of the data.
        - Min: The smallest value in the column.
        - 25% (First Quartile, Q1): The value below which 25% of the data falls.
        - 50% (Median, Q2): The middle value of the data (the second quartile).
        - 75% (Third Quartile, Q3): The value below which 75% of the data falls.
        - Max: The largest value in the column.

[Chatbot transcript](https://chatgpt.com/share/bec27757-3bef-4001-838e-e8d29383979c)

## Question 7

1. When you want to remove rows that contain missig values rather than deleting an entire column, `df.dropna()` would be preferred over `del df['col']`.

    Example:



In [1]:
import pandas as pd
import numpy as np

data = {'Name': ['Alice', 'Bob', 'Charlie', np.nan],
        'Age': [25, np.nan, 35, 28],
        'City': ['New York', 'Los Angeles', np.nan, 'Chicago'],
        'Score': [85, 78, np.nan, 92]}

df = pd.DataFrame(data)
print(df)

      Name   Age         City  Score
0    Alice  25.0     New York   85.0
1      Bob   NaN  Los Angeles   78.0
2  Charlie  35.0          NaN    NaN
3      NaN  28.0      Chicago   92.0


- This produces a DataFrame. 

   We want to remove rows where any missing values(NaN) are present because we do not want incomplete data. In this case, `df.dropna()` removes only the rows with missing values. If we were to use `del df['col']`, we would completely remove a column, which may not be desirable if we only want to handle missing data without deleting useful columns.

In [2]:
# Drop rows where any value is NaN
df_cleaned = df.dropna()
print(df_cleaned)

    Name   Age      City  Score
0  Alice  25.0  New York   85.0


- Only the first row remains because it is the only one without any missing values.

2. When you want to remove an entire column that is no longer needed, regardless of whether it contains missing values or not, `del df['col']` would be preferred over `df.dropna()`. 

    Example:

In [3]:
import pandas as pd
import numpy as np

data = {'Name': ['Alice', 'Bob', 'Charlie', 'David'],
        'Age': [25, np.nan, 35, 28],
        'City': ['New York', 'Los Angeles', 'Chicago', 'Boston'],
        'Score': [85, 78, np.nan, 92],
        'Notes': ['Good performance', 'Late submission', np.nan, 'Excellent']}
df = pd.DataFrame(data)
print(df)

      Name   Age         City  Score             Notes
0    Alice  25.0     New York   85.0  Good performance
1      Bob   NaN  Los Angeles   78.0   Late submission
2  Charlie  35.0      Chicago    NaN               NaN
3    David  28.0       Boston   92.0         Excellent


- For instance, there is a "Notes" column that contains comments or metadata which is irrelevant to the analysis, so we want to remove it entirely. Regardless of whether it contains missing values or not, we just want to remove it, so we would use `del df['col']` to remove that whole column.

In [4]:
# Delete the 'Notes' column
del df['Notes']
print(df)

      Name   Age         City  Score
0    Alice  25.0     New York   85.0
1      Bob   NaN  Los Angeles   78.0
2  Charlie  35.0      Chicago    NaN
3    David  28.0       Boston   92.0


3. Applying `del df['col']` before `df.dropna()` is important because removing an unnecessary column first prevents the accidental loss of valuable data when removing rows with missing values. 

    Example: 
    
    We have a DataFrame where one column (Notes) contains irrelevant information, but it has NaN values. Some rows in other columns also contain missing values, and we'd like to remove those rows after deleting the irrelevant column. 

In [1]:
import pandas as pd
import numpy as np

data = {'Name': ['Alice', 'Bob', 'Charlie', 'David'],
        'Age': [25, np.nan, 35, 28],
        'City': ['New York', 'Los Angeles', 'Chicago', 'Boston'],
        'Score': [85, 78, np.nan, 92],
        'Notes': [np.nan, 'Late submission', np.nan, 'Excellent']}
df = pd.DataFrame(data)
print("Original DataFrame:")
print(df)


Original DataFrame:
      Name   Age         City  Score            Notes
0    Alice  25.0     New York   85.0              NaN
1      Bob   NaN  Los Angeles   78.0  Late submission
2  Charlie  35.0      Chicago    NaN              NaN
3    David  28.0       Boston   92.0        Excellent


- Since if we remove the rows that has missing values first, some useful row might get deleted too if they have a missing value on the irrelevant columns. So we should use `del df['col']` first, to delete the unnecessary Notes column before applying `df.dropna()` to remove rows with missing values in the remaining columns.

In [2]:
# Step 1: Delete the 'Notes' column first
del df['Notes']
print("\nAfter deleting 'Notes' column:")
print(df)

# Step 2: Now, apply dropna to remove rows with any missing values
df_cleaned = df.dropna()
print("\nAfter applying dropna():")
print(df_cleaned)


After deleting 'Notes' column:
      Name   Age         City  Score
0    Alice  25.0     New York   85.0
1      Bob   NaN  Los Angeles   78.0
2  Charlie  35.0      Chicago    NaN
3    David  28.0       Boston   92.0

After applying dropna():
    Name   Age      City  Score
0  Alice  25.0  New York   85.0
3  David  28.0    Boston   92.0


- Rows with missing values in the Age or Score columns are removed after deleting the Notes column. The important part is that we didn't lose any data because of missing values in Notes since we already removed that column.

If we use `df.dropna()` before `del df['col']`:

In [7]:
import pandas as pd
import numpy as np

data = {'Name': ['Alice', 'Bob', 'Charlie', 'David'],
        'Age': [25, np.nan, 35, 28],
        'City': ['New York', 'Los Angeles', 'Chicago', 'Boston'],
        'Score': [85, 78, np.nan, 92],
        'Notes': [np.nan, 'Late submission', np.nan, 'Excellent']}
df = pd.DataFrame(data)
# Step 1: Apply dropna to remove rows with any missing values
df_cleaned = df.dropna()
print("\nAfter applying dropna() before deleting 'Notes':")
print(df_cleaned)

# Step 2: Now, delete the 'Notes' column
del df_cleaned['Notes']
print("\nAfter deleting 'Notes' column:")
print(df_cleaned)


After applying dropna() before deleting 'Notes':
    Name   Age    City  Score      Notes
3  David  28.0  Boston   92.0  Excellent

After deleting 'Notes' column:
    Name   Age    City  Score
3  David  28.0  Boston   92.0


- In this case, only one row remains because by applying `df.dropna()` first, rows with missing values in any column, including the Notes column were removed. Even though we didn’t need the Notes column, applying it first caused rows with missing values in Notes, such as Alice's row to be removed, which led to unnecessary data loss in other columns like Age, City, and Score.

4. 

   

In [2]:
import pandas as pd
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
titanic_data = pd.read_csv(url)
titanic_data


PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

I want to know the number of people who survived in this dataset and have complete personal imformation, Cabin, Fare and Ticket columns are irrelevant imformation.

In [3]:
df = titanic_data
del df['Cabin']
del df['Ticket']
del df['Fare']
print('After deleting irrelevant columns')
df

After deleting irrelevant columns


PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch Embarked  
0        0        S  
1        0        C  
2        0        S  
3        0        S  
4        0        S  
..     ...      ...  
886      0        S  
887      0        S  
888      2        S  
889      0        C  
890      0        Q  

[891 rows x 9 columns]

In [4]:
df_cleaned = df.dropna()
print('After removing people with missing imformations')
df_cleaned

After removing people with missing imformations


PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
885          886         0       3   
886          887         0       2   
887          888         1       1   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch Embarked  
0        0        S  
1        0        C  
2        0        S  
3        0        S  
4        0        S  
..     ...      ...  
885      5        Q  
886      0        S  
887      0        S  
889      0        C  
890      0        Q  

[712 rows x 9 columns]

In [5]:
# Remove rows where 'Survived' is 0
df_cleaned = df[df['Survived'] != 0]
print('People survived with complete personal imformation')
df_cleaned

People survived with complete personal imformation


PassengerId  Survived  Pclass  \
1              2         1       1   
2              3         1       3   
3              4         1       1   
8              9         1       3   
9             10         1       2   
..           ...       ...     ...   
875          876         1       3   
879          880         1       1   
880          881         1       2   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
..                                                 ...     ...   ...    ...   
875                   Najib, Miss. Adele Kiamie "Jane"  female  15.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
880       Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch Embarked  
1        0        C  
2        0        S  
3        0        S  
8        2        S  
9        0        C  
..     ...      ...  
875      0        C  
879      1        C  
880      1        S  
887      0        S  
889      0        C  

[342 rows x 9 columns]

This would be the Titanic dataset of people who survied and have complete personal imformation. By comparing this dataset to the complete Titanic dataset with no modification, we can see that it now have 342 rows and 9 columns which is less than before (891 rows and 12 columns).

### Summary of Interactions with ChatGPT

1. Handling Missing Data:

- We discussed using `df.dropna()` to remove rows or columns containing missing values (NaN), and `del df['col']` to delete specific columns.
- Example: Using `df.dropna()` is useful when you want to keep all columns and only remove rows with incomplete data. `del df['col']` is more appropriate when a column is irrelevant and should be removed regardless of missing values.

2. Preference Between `df.dropna()` and `del df['col']`:

- Scenario for `df.dropna()`: We remove rows containing any NaN values while retaining the columns.
- Scenario for `del df['col']`: We delete a column entirely, even if it contains missing values, because it’s no longer needed.
- Conclusion: Use `df.dropna()` for dropping rows with missing data across columns and `del df['col']` to remove unnecessary columns.

3. Order of Operations:

- We examined why using `del df['col']` before `df.dropna()` can be important. Deleting irrelevant columns first prevents accidental loss of data due to missing values in columns you intend to delete anyway.
- Example: Deleting a column with NaN values first ensures that `df.dropna()` removes only relevant missing values, reducing unnecessary data loss.

4. Deleting Multiple Columns:

- We clarified that `del df['col']` cannot delete multiple columns at once, but `df.drop(columns=[...])` can.
- Example: To delete multiple columns, we used `df.drop(columns=['col1', 'col2'])`.

5. Handling Zeros Instead of NaNs:

- We discussed that `df.dropna()` cannot be used to remove rows containing zeros since zeros are not considered NaN. Instead, boolean indexing (e.g., `df[df['col'] != 0]`) should be used.
- Alternative: Zeros can be replaced with NaN using `df.replace(0, np.nan)` and then `df.dropna()` can be applied to remove the rows.

[Chatbot transcript](https://chatgpt.com/share/1980943e-eab5-4356-b777-3f4b38cb889d)

## Question 8

1. 

 We first break down `df.groupby("col1")["col2"].describe()`,

- `df.groupby("col1")`:

    This part groups the DataFrame df by the unique values in the column "col1".
    After grouping, operations can be performed on each group separately.
- `["col2"]`:

    After grouping by "col1", this selects the column "col2" from each group. The operation will now focus only on this column for each group.
- `.describe()`:

    The `describe()` method provides a summary of descriptive statistics for the selected column in each group.


The `.describe()` method automatically ignores any missing values in "col2" during these calculations. 


`df.groupby("col1")["col2"].describe()` computes summary statistics for each group in col1 for the values in col2.

Titanic Example:

In [12]:
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df


survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [14]:
# Group by 'sex' and get descriptive statistics for 'age'
grouped_stats = df.groupby("sex")["age"].describe()
grouped_stats

count       mean        std   min   25%   50%   75%   max
sex                                                              
female  261.0  27.915709  14.110146  0.75  18.0  27.0  37.0  63.0
male    453.0  30.726645  14.678201  0.42  21.0  29.0  39.0  80.0

For this example, I grouped the data by 'sex' and get the statistics for 'age' by applying `df.groupby("sex")["age"].describe()`. 

Base on the result one can see that there are 261 females and 453 males (according to the count), and the mean for female is 27.9, for male is 30.7. The information for std, min, 25%, 50%, 75% as well max of female and male are also above. 

2. 

In `df.describe()`, when you have missing values in a dataframe, the count represents the number of non-missing entries for each column. Each column is evaluated independently, so the count varies depending on how many missing values are present for that particular column.

However, when you apply `df.groupby("col1")["col2"].describe()`, the behavior of count changes. As it provides a summary (describe) for 'col2' after grouping by the values in 'col1'. The count shows non-missing values of 'col2' within each group defined by 'col1'. The result reflects how many valid (non-missing) values of 'col2' exist for each value in 'col1'. Unlike `df.describe()`, here the count refers specifically to the size of the group with respect to 'col2' after the grouping operation.


3.

   **A.** Forget to include `import pandas as pd` in your code

In [2]:
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df

NameError: name 'pd' is not defined

Error: `NameError: name 'pd' is not defined`

   For this error, I found the correct way to fix it using both methods, but using a chatbot (ChatGPT) might be quicker and more efficient, by asking questions it provide you with direct interation which gives a faster clarification than reading detailed discussion on websites.
   
   Google: [solution](https://stackoverflow.com/questions/40407090/nameerror-name-pd-is-not-defined)

**B.** Mistype "titanic.csv" as "titanics.csv"

In [3]:
import pandas as pd
url = "titanics.csv"
df = pd.read_csv(url)
df

FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv'

Error: `FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv'`

In [4]:
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanics.csv"
df = pd.read_csv(url)
df

HTTPError: HTTP Error 404: Not Found

Error: `HTTPError: HTTP Error 404: Not Found`

In [5]:
url = "https://titanics.csv"
df = pd.read_csv(url)
df

URLError: <urlopen error [Errno -2] Name or service not known>

Error: `URLError: <urlopen error [Errno -2] Name or service not known>`

 If you have chat with the chatbot before and have given the correct URL to it, it will provide you with the correct URL, which I think will make the correction faster. Besides that working with ChatGPT seems to be the faster in general and is a more convenient option because the error is straightforward, and the solution is provided immediately in context. 

For the three errors I created above the chatbot (ChatGPT) identified the problem of URL and ask me to check the URL, it also provide me with the correct URL from our chat history. 

By using google I found the same error that others faced and the solution for their cases but it shows what should be fixed for our example, but I think for some of them the name of the error already made it clear what's the problem.

Google: [FileNotFoundError](https://www.educative.io/answers/how-to-fix-filenotfounderror-errno-2-no-such-file-or-directory)
        [URLError](https://stackoverflow.com/questions/45744991/urllib2-urlerror-urlopen-error-errno-8)
        [404](https://stackoverflow.com/questions/42441211/python-urllib-error-httperror-http-error-404-not-found)

**C.** Try to use a dataframe before it's been assigned into the variable

In [6]:
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
Df

NameError: name 'Df' is not defined

Error: `NameError: name 'Df' is not defined`


I can not quickly find the solution on internet just by searching the error `NameError: name 'Df' is not defined`, but when using the ChatBot(ChatGPT), it identified the problem immediately. 

**D.** Forget one of the parentheses somewhere the code

In [7]:
df = pd.read_csv(url
df

SyntaxError: '(' was never closed (4270852414.py, line 1)

Error: `SyntaxError: '(' was never closed`

Using a ChatBot (ChatGPT) is faster.

Using google I only found similar SyntaxError and the way to solve them.

Google: [SyntaxError](https://realpython.com/invalid-syntax-python/#missing-parentheses-brackets-and-quotes)

**E.** Mistype one of the names of the chained functions with the code

In [11]:
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv" # we are still using the dataset in Q1
df = pd.read_csv(url)

missing_values = df['Age'].valuecounts(dropna=False)
print(missing_values)

AttributeError: 'Series' object has no attribute 'valuecounts'

Error: `AttributeError: 'Series' object has no attribute 'valuecounts'`

Both methods can solve the problem, but using a ChatBot (ChatGPT) is faster. It quickly pointed out it’s likely that I intended to use the `.value_counts()`.

Google: [Solution](https://stackoverflow.com/questions/54557423/series-object-has-no-attribute-values-counts)

**F.** Use a column name that's not in your data for the groupby and column selection

In [13]:
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

grouped_stats = df.groupby("Sex")["age"].describe()
grouped_stats

KeyError: 'Sex'

Error: `KeyError: 'Sex'`

In [14]:
grouped_stats = df.groupby("sex")["Age"].describe()
grouped_stats

KeyError: 'Column not found: Age'

Error: `KeyError: 'Column not found: Age'`

Both methods can identify the problem which the column does not exist, but the ChatBot (ChatGPT) pointed out that I should use 'age' rather than 'Age' and 'sex' rather than 'Sex'.

Google: [KeyError](https://saturncloud.io/blog/python-pandas-keyerror-none-of-index-are-in-the-columns/#:~:text=One%20of%20the%20most%20common,you%20will%20get%20a%20KeyError.)

**G.** Forget to put the column name as a string in quotes for the groupby and column selection

In [16]:
grouped_stats = df.groupby("sex")[age].describe()
grouped_stats

NameError: name 'age' is not defined

Error: `NameError: name 'age' is not defined`

I can not find a solution on google, but with my previous history of question 8, ChatGPT is capable of identifying the problem, and providing the solution. It also provided me with the correct code.

However, when I opened a new session and only provided ChatGPT with the error `NameError: name 'age' is not defined`, it did not give me the correct solution to fix the problem. Later, when I gave it the code I wrote, and the error, it was able to fix it.


Overall, I think it is easier to work in a ChatBot session to fix the errors, since proven by the examples above, Chatbots is typically faster and more effective in providing an immediate solution and correct the mistake in real-time with the help of memories that you had provided it in the session. A Google search often gave me an overwhelming number of irrelevant posts from users dealing with issues that were only loosely related to mine, even though the error messages were the same. You can find out the problem with it too, but it is time-consuming to find a similar case.

### Summary of Interactions with ChatGPT

1. Initial Analysis on Titanic Dataset:

- You were performing initial simple summary analyses on the Titanic dataset.
- You sought help understanding code such as `df.groupby("col1")["col2"].describe()`.

2. Error Resolution and Code Assistance:

- `NameError: name 'pd' is not defined`:

    - Resolved by ensuring `pandas` was imported with `import pandas as pd`.

- `FileNotFoundError: [Errno 2] No such file or directory`:

    - Addressed by verifying the file path or URL. Suggested using the correct URL for the Titanic dataset: "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv".

- `HTTPError: HTTP Error 404: Not Found`:

    - Resolved by confirming the URL is correct and accessible. If issues persist, try accessing the dataset from a different source or local file.

- `URLError: <urlopen error [Errno -2] Name or service not known>`:

    - Suggested checking the internet connection, URL correctness, and network settings.

- `NameError: name 'Df' is not defined`:

    - Fixed by ensuring consistent use of the correct variable name `df` (case-sensitive).

- `SyntaxError: '(' was never closed`:

    - Addressed by checking for balanced parentheses in function calls and expressions.

- `AttributeError: 'Series' object has no attribute 'valuecounts'`:

    - Corrected by using the proper method `value_counts()`.

- `KeyError: 'Column not found: Age'`:

    - Fixed by verifying column names, ensuring case sensitivity, and stripping any extra whitespace.

- `NameError: name 'age' is not defined`:

    - Resolved by using column names as strings (e.g., "age") and ensuring column names are correctly referenced.


[Chatbot transcript](https://chatgpt.com/share/14404e1f-395e-41cf-a655-a1da1b44c57f)

## Question 9

**Yes!**